### PR TITLE
accelerators: (snax streamers) add streamer flags

### PR DESCRIPTION
**Pull request by jorendumoulin**
This PR extends the streamer configuration attributes so that each streamer dimension can have a flag assigned to it.
This allows us to express the irrelevance of certain dimensions or possible reuse within the accelerator, such that we can program the streamers correctly.

Additionally, this pr also deprecates and starts moving away from the global `spatial_dim` and `temporal_dim` functions. These are only valid when the temporal/spatial dim is equal for every operand, which is only the case for very simple examples.

### CODE REVIEW

File: compiler/transforms/stream_snaxify.py [hunk 1]
```diff
@@ -130,12 +130,6 @@ def generate_one_list(n: int, i: int):
 
             spat_dim = streamer_config.data.spatial_dim()
 
-            # extremely dirty fix:
-            # FIXME: this only works because gemm is the only one with
-            # two spat_dims. This must be fixed with some "virtual" spatial dim
-            if spat_dim == 2:
-                spat_dim = 3
-
```
Comment (jorendumoulin) [inline compiler/transforms/stream_snaxify.py:138]: Solving this is the main takeaway for this PR. If a spatial dimension is irrelevant to a certain operand, the streamer does not exist and should not be programmed. With the streamer flags this is now properly handled.

File: compiler/accelerators/streamers/streamers.py [hunk 2]
```diff
@@ -8,19 +10,83 @@ class StreamerType(StrEnum):
...
+            (StreamerFlag.Normal,) * spatial_dim,
+        )
+
+    @property
+    def temporal_dim(self):
+        return len(self.temporal_dims)
+
+    @property
+    def spatial_dim(self):
+        return len(self.spatial_dims)
```
Comment (jorendumoulin) [inline compiler/accelerators/streamers/streamers.py:89]: The temporal and spatial dim for the streamers of a given operand is still valid.

File: compiler/accelerators/streamers/streamers.py [hunk 3]
```diff
@@ -42,6 +108,7 @@ def size(self) -> int:
         """
         return len(self.streamers)
 
+    @deprecated("Please do not use this function, it is only valid in trivial cases")
     def temporal_dim(self) -> int:
         """
```
Comment (jorendumoulin) [inline compiler/accelerators/streamers/streamers.py:113]: These functions try to determine a single temporal dim for the streamers of several operands, but can only be done if they are all equal. This is often not the case.

Comment (jorendumoulin) [reply compiler/accelerators/streamers/streamers.py:113]: However, my mlperf dev branch still uses this, so I am not removing it just yet to make merging them more graceful.

**Review by JosseVanDelm — APPROVED** [review]
Impressive!

**Pull request merged**